File: myLeetCode/Linked List/24. Swap Nodes in Pairs.cpp
// 24. Swap Nodes in Pairs
// https://leetcode.com/problems/swap-nodes-in-pairs/

#include <iostream>
#include <vector>

struct ListNode {
    int val;
    ListNode *next;

    ListNode(int val) : val(val), next(nullptr) {}
    ListNode(int val, ListNode *next) : val(val), next(next) {}
};

class Solution {
  public:
    ListNode *swapPairs(ListNode *head) {
        ListNode *dummyHead = new ListNode(0);
        dummyHead->next = head;
        ListNode *temp = dummyHead;
        while (temp->next != nullptr && temp->next->next != nullptr) {
            ListNode *node1 = temp->next;
            ListNode *node2 = temp->next->next;
            temp->next = node2;
            node1->next = node2->next;
            node2->next = node1;
            temp = node1;
        }
        ListNode *newHead = dummyHead->next;
        delete dummyHead; // 释放dummyHead内存
        return newHead;
    }
};

int main() {
    std::vector<int> values = {1, 2, 3, 4, 5, 6};
    ListNode *head = nullptr;
    ListNode *curr = nullptr;

    for (int val : values) {
        if (head == nullptr) {
            head = new ListNode(val);
            curr = head;
        } else {
            curr->next = new ListNode(val);
            curr = curr->next;
        }
    }

    Solution sol;
    ListNode *swapped = sol.swapPairs(head);
    while (swapped) {
        std::cout << swapped->val << " ";
        swapped = swapped->next;
    }
    std::cout << std::endl;

    // 释放链表内存
    ListNode *currNode = swapped;
    while (currNode) {
        ListNode *nextNode = currNode->next;
        delete currNode;
        currNode = nextNode;
    }

    return 0;
}

File: myLeetCode/Linked List/24. Swap Nodes in Pairs.py
# 24. Swap Nodes in Pairs
# https://leetcode.com/problems/swap-nodes-in-pairs/

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        dummyHead = ListNode(0)
        dummyHead.next = head
        temp = dummyHead
        while temp.next and temp.next.next:
            node1 = temp.next
            node2 = temp.next.next
            temp.next = node2
            node1.next = node2.next
            node2.next = node1
            temp = node1
        newHead = dummyHead.next
        del dummyHead  # 释放dummyHead内存
        return newHead

def main():
    values = [1, 2, 3, 4, 5, 6]
    head = None
    curr = None
    for val in values:
        if head is None:
            head = ListNode(val)
            curr = head
        else:
            curr.next = ListNode(val)
            curr = curr.next

    # 调用函数进行节点交换
    sol = Solution()
    swapped = sol.swapPairs(head)

    # 输出结果
    while swapped:
        print(swapped.val, end=" ")
        swapped = swapped.next
    print()

if __name__ == "__main__":
    main()